src/node_file.cc

```diff
@@ -93,6 +92,40 @@ class FSReqWrap: public ReqWrap<uv_fs_t> {
 
   size_t self_size() const override { return sizeof(*this); }
 
+  void Resolve(Environment* env, Local<Value> arg) {
+    Local<Context> context = env->context();
+    Local<Value> promise =
+        object()->Get(context, env->oncomplete_string()).ToLocalChecked();
+    if (promise->IsPromise()) {
+      if (promise.As<Promise>()->State() != Promise::kPending) return;
+      Local<Promise::Resolver> resolver = promise.As<Promise::Resolver>();
+      resolver->Resolve(context, arg).FromJust();
```

Comment:
Due to https://github.com/nodejs/node/issues/5691 you’ll want an `InternalCallbackScope` around this … ideally move the class definition to `node_internals.h` in the same way it’s done in https://github.com/nodejs/node/pull/15428 to avoid merge conflicts :)

Reply:
ah yes, thanks for the reminder :-)

Reply:
fwiw, how and where to use `InternalCallbackScope` is about as clear as mud right now :-( ... any chance of having some docs written up on it? Even if only in the form of some code comments.

Reply:
Yea, sure. From `node.h`:

```
 * `MakeCallback()` is a wrapper around this class as well as
 * `Function::Call()`. Either one of these mechanisms needs to be used for
 * top-level calls into JavaScript (i.e. without any existing JS stack).
```

do you have any specific suggestions for how to improve that? should resolving promise state be mentioned explicitly?

Reply:
I'll work up some language.